Source/Urho3D/Input/InputMap.h

```diff
@@ -0,0 +1,234 @@
+    ea::fixed_vector<ControllerAxisMapping, 1> controllerAxes_;
+    ea::fixed_vector<ControllerHatMapping, 1> controllerHats_;
+    ea::fixed_vector<ScreenButtonMapping, 1> screenButtons_;
+
+    /// Serialize content from/to archive. May throw ArchiveException.
+    void SerializeInBlock(Archive& archive);
+    /// Evaluate action state based on current input.
+    float Evaluate(Input* input, UI* ui, float deadZone, int ignoreJoystickId) const;
+};
+}
+
+
+class URHO3D_API InputMap : public SimpleResource
+{
+    URHO3D_OBJECT(InputMap, SimpleResource);
+
+public:
+    const float DEFAULT_DEADZONE{0.1f};
+
+    /// Construct.
+    explicit InputMap(Context* context);
+    ~InputMap() override;
+
+    /// Register object factory and attributes.
+    static void RegisterObject(Context* context);
+
+    /// Force override of SerializeInBlock.
+    void SerializeInBlock(Archive& archive) override;
+
+    /// Set dead zone half-width to mitigate axis drift.
+    void SetDeadZone(float deadZone);
+    /// Get dead zone half-width.
+    float GetDeadZone() const { return deadZone_; }
+
+
+    /// Map keyboard key to the action.
+    void MapKeyboardKey(const ea::string& action, Scancode scancode);
+
+    /// Map controller button to the action.
+    void MapControllerButton(const ea::string& action, ControllerButton button);
+
+    /// Map generic joystick button to the action.
+    void MapJoystickButton(const ea::string& action, unsigned buttonIndex);
+
+    /// Map controller axis to the action.
+    void MapControllerAxis(const ea::string& action, ControllerAxis axis, float neutral, float pressed);
+
+    /// Map generic joystick axis to the action.
+    void MapJoystickAxis(const ea::string& action, unsigned axis, float neutral, float pressed);
+
+    /// Map controller hat to the action.
+    void MapHat(const ea::string& action, HatPosition hatPosition);
+
+    /// Map mouse button to the action.
+    void MapMouseButton(const ea::string& action, MouseButton mouseButton);
+
+    /// Map screen button to the action.
+    void MapScreenButton(const ea::string& action, const ea::string& elementName);
+
+    /// Get mapping for the action.
+    const Detail::ActionMapping& GetMapping(const ea::string& action) const;
+
+    /// Add new metadata variable or overwrite old value.
+    void AddMetadata(const ea::string& name, const Variant& value);
+    /// Remove metadata variable.
+    void RemoveMetadata(const ea::string& name);
+    /// Remove all metadata variables.
+    void RemoveAllMetadata();
+    /// Return metadata variable.
+    const Variant& GetMetadata(const ea::string& name) const;
+    /// Return whether the resource has metadata.
+    bool HasMetadata() const;
+
+    /// Get all action mappings.
+    const ea::unordered_map<ea::string, Detail::ActionMapping>& GetMappings() const { return actions_; }
+
+    /// Evaluate action state.
+    float Evaluate(const ea::string& action);
```

Comment:
We should use `StringHash` to identify actions. Same with storing them in `unordered_map`.

Reply:
I thought @eugeneko transitions away from stringhash. I can bring it back if needed.

Reply:
Wait what? First time im hearing it 👀

Reply:
I would keep both string and StringHash overloads while keeping strings inside.